File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafFinder(self, root: Optional[TreeNode], leaf_value):
        if not root:
            return 
        if(not root.left and not root.right):
            leaf_value.append(root.val)
        self.leafFinder(root.left, leaf_value)
        self.leafFinder(root.right, leaf_value)

    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        leaf_value1 = [] 
        leaf_value2 = [] 
        self.leafFinder(root1, leaf_value1)
        self.leafFinder(root2, leaf_value2)
        return leaf_value1==leaf_value2

    

        

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void leafNode(TreeNode* root, vector<int>& leaf){
        if(root==NULL){
         return;   
        }
        if(root->left==NULL && root->right==NULL){
            leaf.push_back(root->val);
        }
        leafNode(root->left, leaf);
        leafNode(root->right, leaf);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> leaf1;
        vector<int> leaf2;
        leafNode(root1, leaf1);
        leafNode(root2, leaf2);
        return leaf1==leaf2;
    }
};